spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java

```diff
@@ -407,6 +428,216 @@ public void close() throws IOException {
+
+      ClusteredDataWriter<InternalRow> updateWriter = new ClusteredDataWriter<>(
+          writerFactory, dataFileFactory, table.io(),
+          context.dataFileFormat(), context.targetDataFileSize());
+
+      ClusteredPositionDeleteWriter<InternalRow> deleteWriter = new ClusteredPositionDeleteWriter<>(
+          writerFactory, deleteFileFactory, table.io(),
+          context.deleteFileFormat(), context.targetDeleteFileSize());
+
+      this.delegate = new BasePositionDeltaWriter<>(insertWriter, updateWriter, deleteWriter);
+      this.io = table.io();
+      this.dataSpec = table.spec();
+      this.specs = table.specs();
+
+      Types.StructType partitionType = Partitioning.partitionType(table);
+      this.deletePartitionRowWrapper = initPartitionRowWrapper(partitionType);
+      this.deletePartitionProjections = buildPartitionProjections(partitionType, specs);
+
+      this.specIdOrdinal = context.metadataSparkType().fieldIndex(MetadataColumns.SPEC_ID.name());
+      this.partitionOrdinal = context.metadataSparkType().fieldIndex(MetadataColumns.PARTITION_COLUMN_NAME);
+      this.fileOrdinal = context.deleteSparkType().fieldIndex(MetadataColumns.FILE_PATH.name());
+      this.positionOrdinal = context.deleteSparkType().fieldIndex(MetadataColumns.ROW_POSITION.name());
+    }
+
+    @Override
+    public void delete(InternalRow meta, InternalRow id) throws IOException {
+      int specId = meta.getInt(specIdOrdinal);
+      PartitionSpec spec = specs.get(specId);
+
+      InternalRow partition = meta.getStruct(partitionOrdinal, deletePartitionRowWrapper.size());
+      StructProjection partitionProjection = deletePartitionProjections.get(specId);
+      partitionProjection.wrap(deletePartitionRowWrapper.wrap(partition));
+
+      String file = id.getString(fileOrdinal);
+      long position = id.getLong(positionOrdinal);
+      delegate.delete(file, position, spec, partitionProjection);
+    }
+
+    @Override
+    public void update(InternalRow meta, InternalRow id, InternalRow row) throws IOException {
+      delete(meta, id);
+      delegate.update(row, dataSpec, null);
+    }
+
+    @Override
+    public void insert(InternalRow row) throws IOException {
+      delegate.insert(row, dataSpec, null);
+    }
+
+    @Override
+    public WriterCommitMessage commit() throws IOException {
+      close();
+
+      WriteResult result = delegate.result();
+      return new DeltaTaskCommit(result);
+    }
+
+    @Override
+    public void abort() throws IOException {
+      close();
+
+      WriteResult result = delegate.result();
+      cleanFiles(io, Arrays.asList(result.dataFiles()));
+      cleanFiles(io, Arrays.asList(result.deleteFiles()));
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (!closed) {
+        delegate.close();
+        this.closed = true;
+      }
+    }
+  }
+
+  private static class PartitionedDeltaWriter extends BaseDeltaWriter {
```

Comment:
It definitely looks to me like there is a lot of duplication between the partitioned and unpartitioned cases. How bad is it to merge the two as you commented above? I think I would probably prefer that.

Reply:
I combined these two classes in one and exposed a protected field in the parent instead.